index.js

```diff
@@ -48,7 +48,12 @@ CsvWriteStream.prototype._compile = function(headers) {
 CsvWriteStream.prototype._transform = function(row, enc, cb) {
   var isArray = Array.isArray(row)
 
-  if (!isArray && !this.headers) this.headers = Object.keys(row)
+  if (!this.headers) {
+    this.headers = Object.keys(row)
```

Comment:
This seems to be required for generated functions (see `this._arrRow`) to work. To bypass the headers requirement I think we would have to rewrite the function generation.

But I might be wrong I didn't look into the code much - I just made the fix for our use case in hope it might be useful for others too...